github/repos.go

```diff
@@ -534,24 +535,53 @@ type RequiredStatusChecks struct {
 
 // PullRequestReviewsEnforcement represents the pull request reviews enforcement of a protected branch.
 type PullRequestReviewsEnforcement struct {
-	DismissalRestrictions *DismissalRestrictions `json:"dismissal_restrictions"`
-	DismissStaleReviews   bool                   `json:"dismiss_stale_reviews"`
+	// Specifies which users and teams can dismiss pull requets reviews.
+	DismissalRestrictions DismissalRestrictions `json:"dismissal_restrictions"`
+	// Specifies if approved reviews are dismissed automatically, when a new commit is pushed.
+	DismissStaleReviews bool `json:"dismiss_stale_reviews"`
 }
 
 // PullRequestReviewsEnforcementRequest represents request to set the pull request review
 // enforcement of a protected branch. It is separate from PullRequestReviewsEnforcement above
 // because the request structure is different from the response structure.
 type PullRequestReviewsEnforcementRequest struct {
+	// Specifies which users and teams should be allowed to dismiss pull requets reviews. Can be nil to disable the restrictions.
 	DismissalRestrictionsRequest *DismissalRestrictionsRequest `json:"dismissal_restrictions"`
-	DismissStaleReviews          bool                          `json:"dismiss_stale_reviews"`
-}
-
-// PullRequestReviewsEnforcementPatchRequest represents request to patch the pull request review
-// enforcement of a protected branch. It is separate from PullRequestReviewsEnforcementPatchRequest above
+	// Specifies if approved reviews can be dismissed automatically, when a new commit is pushed. (Required)
+	DismissStaleReviews bool `json:"dismiss_stale_reviews"`
+}
+
+// MarshalJSON implements the json.Marshaler interface.
+// Converts nil value of PullRequestReviewsEnforcementRequest.DismissalRestrictionsRequest to empty array
+func (req PullRequestReviewsEnforcementRequest) MarshalJSON() ([]byte, error) {
+	if req.DismissalRestrictionsRequest == nil {
+		newReq := struct {
+			R []bool `json:"dismissal_restrictions"`
+			D bool   `json:"dismiss_stale_reviews"`
+		}{
+			R: []bool{},
+			D: req.DismissStaleReviews,
+		}
+		return json.Marshal(newReq)
+	}
+	newReq := struct {
+		R *DismissalRestrictionsRequest `json:"dismissal_restrictions"`
+		D bool                          `json:"dismiss_stale_reviews"`
+	}{
+		R: req.DismissalRestrictionsRequest,
+		D: req.DismissStaleReviews,
+	}
+	return json.Marshal(newReq)
+}
```

Comment:
This is a really cool solution to work around the GitHub API weirdness. Nice idea! 👍